Name: Shuren He
Course: CSC 44800 Artificial Intelligence
Assignment 4: Judging Flowers

# Part 1: Load the dataset

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/SHUREN55/AI-assignment2/blob/main/AL_assignment4.ipynb)

In [ ]:
#Imports and pip installations (if needed)
import numpy as np

import pandas as pd
from sklearn.datasets import load_iris

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import mean_squared_error, r2_score

from sklearn import svm
from sklearn.svm import SVC
from sklearn import datasets

from sklearn.neural_network import MLPClassifier

In [ ]:
# Load the dataset (load remotely, not locally)
iris = load_iris()
iris_data_frame = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])

In [ ]:
# Output the first 15 rows of the data:
iris_data_frame.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0  
5      0.0  
6      0.0  
7      0.0  
8      0.0  
9      0.0  
10     0.0  
11     0.0  
12     0.0  
13     0.0  
14     0.0

In [ ]:
iris_data_frame.dtypes

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target               float64
dtype: object

In [ ]:
iris_data_frame.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

About the dataset

As we expained previous, there are three species of iris.What are your features and labels? the feature X should be iris['data'] contains sepal length (cm), sepal width (cm), 
petal length (cm) and petal width (cm). The label y will be the target, the number indicates iris class.

# Part 2: Split the dataset into train and test

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = np.array(iris_data_frame.drop(['target'], 1))
y = np.array(iris_data_frame['target'])

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# numpy.random. RandomState() This is a pseudorandom number generator.
# we can also make random_state be a exactly number, so we can control the model we get.
rng = np.random.RandomState(0)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=rng)

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
LR_model = LogisticRegression()
#Training the model
LR_model.fit(X_train, y_train)

LogisticRegression()

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
# We can test our model by the first point in test set
# the first predict_proba() return estimates for all classes are ordered by the label of classes.
# since we have three labal, our return will be a array have three numbers
probabilities = LR_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])

For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 8.233431065579563e-05 |class 1(Versicolour): 0.050494555412413863 |class 2(Virginica): 0.9494231102769304


In [ ]:
# iii. Report on the score for Logistic regression model
LR_confidence = LR_model.score(X_test, y_test)
print('The score of our Logistic Regression Model is',LR_confidence)

The score of our Logistic Regression Model is 1.0


the score of our Logistic Regression Model is perfectly 1, which mean our model always correctly predict the right class 
of iris 

In [ ]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
# The coefficients and intercepts:
print('Coefficients: \n', LR_model.coef_)
print('Intercepts: \n', LR_model.intercept_)

Coefficients: 
 [[-0.41087752  0.92630564 -2.43464541 -1.04770048]
 [ 0.40315384 -0.32622858 -0.11363271 -0.88710626]
 [ 0.00772368 -0.60007705  2.54827812  1.93480674]]
Intercepts: 
 [  9.46338834   2.4453442  -11.90873254]


There are three boundary line for each class. At above, the first list the cofficicent list is the cofficients of the first boundary line for class 0.
 Each conefficient relatives to one features, we have four features which perfectly match the coefficients. And the intercepts of the first boundary line is 9.4634.
The other two boundary line follow the same format.

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVC_model = SVC(kernel = 'linear', random_state = rng, probability = True)
SVC_model.fit(X_train, y_train)

SVC(kernel='linear', probability=True,
    random_state=RandomState(MT19937) at 0x19C02240640)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
probabilities = SVC_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])

For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 0.009125973292447928 |class 1(Versicolour): 0.007037937656916896 |class 2(Virginica): 0.9838360890506351


In [ ]:
# iii. Report on the score for the SVM, what does the score measure?
SVC_confidence = SVC_model.score(X_test, y_test)
print('The score of our Support Vector Classifier is',SVC_confidence)

The score of our Support Vector Classifier is 1.0


If we compare the second part, the probalities of each possible class of the same is very similarly to the last we got from Logistic regression model.
The class 3 has the highest ratio.

# Part 5: Neural Network

In [ ]:
NN_model = MLPClassifier(solver='lbfgs',tol = 1e-6,hidden_layer_sizes=(5,4),activation="logistic",alpha=1e-1,max_iter=1000,random_state=rng)
NN_model.fit(X_train, y_train)

MLPClassifier(activation='logistic', alpha=0.1, hidden_layer_sizes=(5, 4),
              max_iter=1000, random_state=RandomState(MT19937) at 0x19C02240640,
              solver='lbfgs', tol=1e-06)

In [ ]:
# ii. For a sample datapoint, predict the probabilities for each possible class
probabilities = NN_model.predict_proba(X_test[:1, :])
print('For the iris sample:', X_test[:1, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])

For the iris sample: [[5.8 2.8 5.1 2.4]]
class 0(Setosa): 0.0014382310357031697 |class 1(Versicolour): 0.004548053252215499 |class 2(Virginica): 0.9940137157120813


In [ ]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN_confidence = NN_model.score(X_test, y_test)
print('The score of our Neutral Network Model is',NN_confidence)

The score of our Neutral Network Model is 1.0


# Part 6: K-Nearest Neighbors

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

probabilities = knn.predict_proba(X_test[10:11, :])
print('For the iris sample:', X_test[10:11, :])
print('class 0(Setosa):',probabilities[0][0],'|class 1(Versicolour):',probabilities[0][1],'|class 2(Virginica):',probabilities[0][2])

# iii. Report on the score for kNN, what does the score measure?
confidence = knn.score(X_test, y_test)
print('The score of our K-Nearest Model is',confidence)

For the iris sample: [[6.1 2.6 5.6 1.4]]
class 0(Setosa): 0.0 |class 1(Versicolour): 0.2 |class 2(Virginica): 0.8
The score of our K-Nearest Model is 1.0


The K-Nearest Neighbors models will define the class of the point by nearest n point. It is also a good idea to prediect new point give to 
us. 

# Part 7: Conclusions and takeaways

Base on the previous work, the all scores of today's meachine model is 1, which means all models are accurated. The building of neutal network model, we gradually get closer to optimal answer. K-NN algorithm is simple to understand and equally easy to implement. If one class only get few input data, then the model will ultimately give a lot of preference to majority set. This might result in getting the less common class wrongly classified.